### PR TITLE
Pdf print

### PR DESCRIPTION
**Pull request by darenegade**
---
name: Pull request
about: Contribute your changes to the project
title: '[pull-request]'
labels: ''
assignees: ''
---
**Description**
Add new Consumer to accept PDF prints via S3

**Reference**
Issues #XXX

**Checklist**

- [x] Acceptance criteria are met
- [x] Pipeline has been run successfully
- [x] JUnit tests have been run successfully
- [ ] Changelog is adjusted
- [ ] [Documentations](https://git.muenchen.de/groups/digitalisierung/-/wikis/Dokumentationen) are completed
- [ ] Frontend is tested
- [ ] Created sub-branches are deleted
- [x] Boards are updated ( [Digitalisierung](https://git.muenchen.de/groups/digitalisierung/-/boards) , [Support](https://git.muenchen.de/digitalisierung/digiwf-support) , [Zenhub](https://app.zenhub.com/workspaces/digiwf-621f70bf50ea1100120b7e93/board) )
- [ ] feature process is created or extended
- [ ] [git-ops](https://git.muenchen.de/digitalisierung/digiwf-ops) is customized
- [ ] Openshift environments are prepared (Secrets, etc.)

### CODE REVIEW

File: digiwf-oce-cosmos-integration-service/src/main/java/io/muenchendigital/digiwf/ocecosmos/integration/api/streaming/OceCosmosStreamingEventListener.java [hunk 1]
```diff
@@ -67,12 +71,73 @@ public Consumer<Message<JobRequestDto>> requestPrint() {
...
+
+            if(printRequest.getS3Files() == null) {
+                this.correlateMessageService.sendCorrelateMessage(
+                        message.getHeaders(),
+                        Map.of(RESPONSE, new OceCosmosErrorDto("Presigned S3 File-URLs needed"))
+                );
+                return;
+            }
+
+            if(printRequest.getS3Files().size() != 1) {
```
Comment (axd4) [inline digiwf-oce-cosmos-integration-service/src/main/java/io/muenchendigital/digiwf/ocecosmos/integration/api/streaming/OceCosmosStreamingEventListener.java:98]: Why use a list if only one file is supported?

File: docs/element-template-s3-print.json [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        "type": "camunda:in",
+        "expression": true,
+        "target": "app_type_name"
+      },
+      "constraints": {
+        "notEmpty": true
+      }
+    },
+    {
+	  "label": "Versandart",
```
Comment (axd4) [inline docs/element-template-s3-print.json:77]: You could implement a dropdown here, if there's only two valid inputs

File: docs/element-template-templatet-print.json [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    },
+    {
+	  "label": "Versandart",
+	  "description": "Versandart des Dokuments (POSTVERSAND, DIENSTVERSAND) eingeben.",
+      "type": "String",
+      "value": "",
+      "binding": {
+        "type": "camunda:in",
+        "expression": true,
+        "target": "BewohnerVersandart"
```
Comment (axd4) [inline docs/element-template-templatet-print.json:69]: BewohnerVersandart => deliveryType?

File: docs/element-template-templatet-print.json [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    },
+	    {
+	  "label": "Abholort",
+	  "description": "Abholort des Druckdokuments nur bei Versandart DIENSTVERSAND eingeben.",
+      "type": "String",
+      "value": "",
+      "binding": {
+        "type": "camunda:in",
+        "expression": true,
+        "target": "AbholortDienstversand"
```
Comment (axd4) [inline docs/element-template-templatet-print.json:83]: => pickupLocation?

I think the other two (druckDaten, parkausweisVorlage) might also be incorrect

**Review by axd4 — COMMENTED** [review]
Other than that, it looks good. I think in the email artifact we generate the presigned urls inside the artifact. Maybe we should change one or the other, so that for the process developers it always looks and acts the same?

**Comment by darenegade** [comment]
> Other than that, it looks good. I think in the email artifact we generate the presigned urls inside the artifact. Maybe we should change one or the other, so that for the process developers it always looks and acts the same?

The s3 integration in the email artifact has changed: https://github.com/it-at-m/digiwf-email-integration/pull/25
I just copied the code from email, so this integration looks and acts the same

**Comment by axd4** [comment]
> > Other than that, it looks good. I think in the email artifact we generate the presigned urls inside the artifact. Maybe we should change one or the other, so that for the process developers it always looks and acts the same?
> 
> The s3 integration in the email artifact has changed: [it-at-m/digiwf-email-integration#25](https://github.com/it-at-m/digiwf-email-integration/pull/25) I just copied the code from email, so this integration looks and acts the same

Ahh, I didn't know that. No need to change this then! :)

File: digiwf-oce-cosmos-integration-service/src/main/java/io/muenchendigital/digiwf/ocecosmos/integration/api/streaming/OceCosmosStreamingEventListener.java (hunk 1 above)
Comment (darenegade) [reply digiwf-oce-cosmos-integration-service/src/main/java/io/muenchendigital/digiwf/ocecosmos/integration/api/streaming/OceCosmosStreamingEventListener.java:98]: The s3 integration always returns an array, so to reduce the js mapping code in the workflow, i accept an array here. Beside that multi file support may be needed in the future

File: docs/element-template-s3-print.json (hunk 2 above)
Comment (darenegade) [reply docs/element-template-s3-print.json:77]: These files are not under my control as they are created by the process designer.

File: docs/element-template-templatet-print.json (hunk 3 above)
Comment (darenegade) [reply docs/element-template-templatet-print.json]: correct template uploaded

File: docs/element-template-templatet-print.json (hunk 4 above)
Comment (darenegade) [reply docs/element-template-templatet-print.json]: correct template uploaded

File: digiwf-oce-cosmos-integration-service/src/main/java/io/muenchendigital/digiwf/ocecosmos/integration/api/configuration/OceCosmosConfiguration.java [hunk 5]
```diff
@@ -17,6 +17,7 @@
 public class OceCosmosConfiguration {
 
     private static final String HEADER_PRINT_REQUEST = "requestPrint";
+    private static final String HEADER_PDF_PRINT_REQUEST = "requestPDFPrint";
```
Comment (lmoesle) [inline digiwf-oce-cosmos-integration-service/src/main/java/io/muenchendigital/digiwf/ocecosmos/integration/api/configuration/OceCosmosConfiguration.java:20]: You could configure this in the application.properties.

e.g.
```
io.muenchendigital.digiwf.streaming.typeMappings.requestPDFPrint=requestPDFPrint
io.muenchendigital.digiwf.streaming.typeMappings.requestPrint=requestPrint
```

See Spring Cloud Stream Utils documentation https://github.com/it-at-m/digiwf-spring-cloudstream-utils#function-routing-configuration (Note: Spring Cloud Stream Utils moved to digiwf-core monorepo but the documenation is not migrated yet).

**Review by lmoesle — APPROVED** [review]
It looks good to me. Well done @darenegade

File: digiwf-oce-cosmos-integration-service/src/main/java/io/muenchendigital/digiwf/ocecosmos/integration/api/configuration/OceCosmosConfiguration.java (hunk 5 above)
Comment (darenegade) [reply digiwf-oce-cosmos-integration-service/src/main/java/io/muenchendigital/digiwf/ocecosmos/integration/api/configuration/OceCosmosConfiguration.java:20]: I may add that in the future. Currently it doesn't have to be configurable

**Pull request merged**